blockchain/v2/reactor.go

```diff
@@ -226,13 +230,56 @@ func (r *BlockchainReactor) SetLogger(logger log.Logger) {
 func (r *BlockchainReactor) Start() error {
 	r.reporter = behaviour.NewSwitchReporter(r.BaseReactor.Switch)
 	if r.fastSync {
-		go r.scheduler.start()
-		go r.processor.start()
-		go r.demux()
+		err := r.startSync(nil)
+		if err != nil {
+			return fmt.Errorf("failed to start fast sync: %w", err)
+		}
 	}
 	return nil
 }
 
+// startSync begins a fast sync, signalled by r.events being non-nil. If state is non-nil,
+// the scheduler and processor is updated with this state on startup.
+func (r *BlockchainReactor) startSync(state *state.State) error {
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+	if r.events != nil {
+		return errors.New("fast sync already in progress")
+	}
+	r.events = make(chan Event, chBufferSize)
+	go r.scheduler.start()
+	go r.processor.start()
+	if state != nil {
+		// We have to wait for the processor and scheduler to start, otherwise they will
+		// ignore these messages. Ideally start() should spawn a goroutine internally,
+		// or somehow signal that it's ready.
+		time.Sleep(100 * time.Millisecond)
```

Comment:
Would we be able to block on ` <-r.scheduler.ready()` and `<-r.processor.ready()` here?

Reply:
Yes, we would. 🤦‍♂️ Thank you!